util/deephash/deephash.go

```diff
@@ -161,26 +162,218 @@ func (h *hasher) hashUint64(i uint64) {
 	binary.LittleEndian.PutUint64(h.scratch[:8], i)
 	h.bw.Write(h.scratch[:8])
 }
+func (h *hasher) hashVarint(x uint64) {
+	i := 0
+	for x >= 0x80 {
+		h.scratch[i] = byte(x) | 0x80
+		x >>= 7
+		i++
+	}
+	h.scratch[i] = byte(x)
+	h.bw.Write(h.scratch[:i+1])
+}
 
 var uint8Type = reflect.TypeOf(byte(0))
 
-func (h *hasher) hashValue(v reflect.Value) {
-	if !v.IsValid() {
-		return
+// returns ok if it was handled; else slow path runs
+type typeHasherFunc func(h *hasher, v reflect.Value) (ok bool)
+
+// A non-nil typeHasherFuture is populated into the typeHasher map
+// when its type is first requested, before its func is created.
+// Its func field fn is only populated once the type has been created.
+// This is used for recursive types.
+type typeHasherFuture struct {
```

Comment:
I think it would be simpler to solve the cyclic-types problem with lazy loading. This avoids this "partial" state problem you have going on here and subtle bugs that can occur as a result (see my comment in `hashValue`).

The lazy-loading approach uses a `sync.Once` that triggers upon the first call to `typeHasherFunc`, which performs the lookup lazily.

Reply:
Right now you have:
```go
for i, n := 0, t.NumField(); i < n; i++ {
	...
	fields = append(fields, indexAndFuture{i, getTypeHasherFutureLocked(sf.Type)})
}
return func(h *hasher, v reflect.Value) bool {
	for _, f := range fields {
		...
	}
	return true
}
```
the partial state handling is to make sure the recursive call to `getTypeHasherFutureLocked` doesn't result in infinite recursive.

I'm proposing it become like:
```go
var once sync.Once
init := func() {
	for i, n := 0, t.NumField(); i < n; i++ {
		...
		fields = append(fields, indexAndFuture{i, getTypeHasherFutureLocked(sf.Type)})
	}
}
return func(h *hasher, v reflect.Value) bool {
	once.Do(init)
	for _, f := range fields {
		...
	}
	return true
}
```
Thus, we don't know have to worry about infinite recursion with `getTypeHasherFutureLocked`.